ext/statsd/statsd.c

```diff
@@ -0,0 +1,376 @@
+  if (st_lookup(builder->normalized_names_cache, key, &val)){
+    return (VALUE)val;
+  } else if (builder->normalized_names_cache->num_entries < NORMALIZED_NAMES_CACHE_MAX) {
+    cached = normalize_name(builder, self, name);
+    st_insert(builder->normalized_names_cache, key, (st_data_t)cached);
+    return cached;
+  }
+  return normalize_name(builder, self, name);
+  RB_GC_GUARD(cached);
+#else
+  return normalize_name(self, name);
+#endif
+}
+
+/* pure function not exposed to ruby with an intermediate bounded cache */
+static VALUE
+normalized_tags_cached(struct datagram_builder *builder, VALUE self, VALUE tags)
+{
+#ifdef NORMALIZED_TAGS_CACHE_ENABLED
+  st_index_t key;
+  st_data_t val;
+  VALUE cached;
+  // More involved hashing as we need to hash on the content of the container too
+  // XXX: revisit
+  key = (st_index_t)(FIX2LONG(rb_hash(tags)));
+  if (st_lookup(builder->normalized_tags_cache, key, &val)){
+    return (VALUE)val;
+  } else if (builder->normalized_tags_cache->num_entries < NORMALIZED_TAGS_CACHE_MAX) {
+    cached = rb_funcall(self, idNormalizeTags, 1, tags);
+    st_insert(builder->normalized_tags_cache, key, (st_data_t)cached);
+    return cached;
+  }
+  return rb_funcall(self, idNormalizeTags, 1, tags);
+  RB_GC_GUARD(cached);
+#else
+  return rb_funcall(self, idNormalizeTags, 1, tags);
+#endif
+}
+
+inline static bool append_normalized_tags(struct datagram_builder *builder, VALUE normalized_tags, int trim_trailing_comma)
+{
+  VALUE tag;
+  int tags_len = 0, chunk_len = 0, i = 0;
+  tags_len = (int)RARRAY_LEN(normalized_tags);
+  for (i = 0; i < tags_len; ++i) {
+    tag = RARRAY_AREF(normalized_tags, i);
+    chunk_len = (int)RSTRING_LEN(tag);
+    if (builder->len + chunk_len > DATAGRAM_SIZE_MAX) return false;
+    memcpy(builder->datagram + builder->len, StringValuePtr(tag), chunk_len);
+    builder->len += chunk_len;
+    if (!trim_trailing_comma || i < tags_len - 1) {
+      if (builder->len + 1 > DATAGRAM_SIZE_MAX) return false;
+      memcpy(builder->datagram + builder->len, ",", 1);
+      builder->len += 1;
+    }
+  }
+  return true;
+}
+
+static VALUE
+generate_generic_datagram(VALUE self, VALUE name, VALUE value, const char *type, VALUE sample_rate, VALUE tags) {
+  VALUE normalized_name, str_value, str_sample_rate;
+  VALUE normalized_tags = Qnil;
+  char sr_buf[SAMPLE_RATE_SIZE_MAX];
+  bool empty_tags = true;
+  long chunk_len = 0;
+  get_datagram_builder_struct(self);
+
+  builder->len = builder->prefix_len;
+
+  if (NIL_P(normalized_name = normalize_name_fast_path(self, name))) {
+    normalized_name = normalized_names_cached(builder, self, name);
+  }
+  chunk_len = RSTRING_LEN(normalized_name);
+  if (builder->len + chunk_len > DATAGRAM_SIZE_MAX) goto finalize_datagram;
+  memcpy(builder->datagram + builder->len, StringValuePtr(normalized_name), chunk_len);
+  builder->len += chunk_len;
+
+  if (builder->len + 1 > DATAGRAM_SIZE_MAX) goto finalize_datagram;
```

Comment:
You could maybe extract this into an inline function.

Reply:
Was thinking about it too, but for flow control to convert the buffer to a Ruby String and return, the goto is local.

A static function would require length and the builder as arguments too and wouldn't make a difference to code size.

Thoughts?

Reply:
Or let overflow checking be a compile time option and allow arbitrary size datagrams if disabled (effectively the current Ruby implementation's behaviour)

Reply:
I was thinking that this would make more sense if we were raising an exception.

Reply:
The risk / regression factor in raising exceptions is that we'd violate a loose contract of "allowing any size datagrams to be built" to "raising an exception that can fan out to the whole host application"

Thoughts on a middle ground that either:

* Implements overflow guards, truncates the datagram and emit a warning with `rb_warn("StatsD datagram truncated to X bytes")`. Easy to provide a reason too, as with the logging pipeline.

* Follow the existing behavior of supporting arbitrary size datagrams: static buffer that can overflow to a heap allocated buffer OR move the buffer to the heap. Both of these aren't difficult, but creates many more branches and complexity too.

?